woocommerce/navigation
# SlotFill POC

As an alternate approach to config based router aware links proposed in https://github.com/woocommerce/woocommerce-admin/pull/5164, this PR explores a slotFill approach.

The benefits to slotFill are that navigation items can be populated from a plugin or wc-admin without WC Navigation having a dependency on wc-admin or any other router-aware plugins.

### Consumption of Fills

Any plugin that would like to add navigational items can do so by simply rendering the Fill housed in available components `WooNavigationMenu` and `WooNavigationItem`

```js
import { registerPlugin } from "@wordpress/plugins";

const { WooNavigationMenu, WooNavigationItem } = window.wc;

const MyPlugin = () => {
  return (
    <>
      <WooNavigationMenu menu="woocommerce">
        <WooNavigationItem
          item="example-1"
          title="Examples"
          navigateToMenu="my-plugin-examples"
        />
      </WooNavigationMenu>
      <WooNavigationMenu menu="my-plugin-examples" parentMenu="woocommerce">
        <WooNavigationItem item="example-1">
          <a href="www.example.com/1">Example 1</a>
        </WooNavigationItem>
        <WooNavigationItem item="example-2">
          <a href="www.example.com/2">Example 2</a>
        </WooNavigationItem>
      </WooNavigationMenu>
    </>
  );
};

registerPlugin("my-plugin", { render: MyPlugin });
```

![Screen Shot 2020-09-23 at 4 32 20 PM](https://user-images.githubusercontent.com/1922453/93967045-66e13100-fdba-11ea-8c86-89dcc7f25d26.png)


### Test instructions

1. In your plugins folder clone, install, and start https://github.com/psealock/examples.
2. Link `@wordpress/components` in this repository.
3. See the Example links functioning in the sidebar.

### Concerns

I really like the declarative nature of slotFill. From the consumer's perspective, its could be very straightforward. Here are some concerns I can think of:

* As mentioned elsewhere, publishing `WooNavigationMenu` and `WooNavigationItem` will require Lerna to be added so that these components can be gathered as an externals. This PR simply places them on the window directly.
* Location, order, and grouping - In this example, an entire Menu is added, but how to add a Slot for a specific parent menu or group within that menu isn't straightforward. And what control over order would we have?
* In theory, this can replace having to register PHP side depending on our route matching solution. Maybe `NavigationContext` can be used to gather a list of `items` for matching.
